File: src/main/java/com/meti/Strings.scala
package com.meti

object Strings {
  def stream(input: String): Stream[Char] = new StreamImpl(input)

  private class StreamImpl(val input: String) extends AbstractStream[Char] {
    private var counter = 0

    @throws[EndOfStreamException]
    override def head: Char = if (counter < input.length) {
      counter += 1
      input.charAt(counter - 1)
    }
    else throw EndOfStreamException("No more characters left.")
  }

}

File: src/main/java/com/meti/Strings.java
package com.meti;

public class Strings {


	static StringStream stream(String input) {
		return new StringStream(input);
	}

	public Strings() {
	}

	static class StringStream extends AbstractStream<Character> {
		private final String input;
		private int counter = 0;

		private StringStream(String input) {
			this.input = input;
		}

		@Override
		public Character head() throws EndOfStreamException {
			if (counter < input.length()) {
				return input.charAt(counter++);
			} else {
				throw new EndOfStreamException("No more characters left.");
			}
		}
	}
}
